index.js

```diff
@@ -14,63 +14,81 @@ const symbolicate = async (options) => {
   const {force, file} = options
   const cacheDirectory = path.join(__dirname, 'cache', 'breakpad_symbols')
 
-  const symbolCache = new Map
   const dumpText = await fs.promises.readFile(file, 'utf8')
   const images = binaryImages(dumpText)
   const electronImage = images.find(v => /electron/i.test(v.library))
   if (electronImage) console.error(`Found Electron ${electronImage.version}`)
   else console.error('No Electron image found')
 
-  let result = []
+  const lines = dumpText.split(/\r?\n/);
 
-  for (const line of dumpText.split(/\r?\n/)) {
+  const linesByImage = new Map();
+  for (const [lineIndex, line] of lines.entries()) {
     const parsedLine = parseAddressLine(line)
-    if (parsedLine) {
-      const library = parsedLine.libraryBaseName || parsedLine.libraryId
-      const image = images.find(i => i.library === library || path.basename(i.path) === library)
-      if (image) {
-        const offset = parsedLine.address - image.startAddress
-        const res = await symbolicateOne({
-          image,
-          offset
-        })
-        if (res) {
-          result.push(line.substr(0, parsedLine.replace.from) + res.func.name + line.substr(parsedLine.replace.from + parsedLine.replace.length))
-          continue
-        }
+    if (!parsedLine) {
+      continue;
+    }
+
+    const library = parsedLine.libraryBaseName || parsedLine.libraryId
+    const image = images.find(i => i.library === library || i.basename === library)
+    if (!image) {
+      continue;
+    }
+
+    const offset = parsedLine.address - image.startAddress
+    const imageKey = `${image.debugId}/${image.basename}`;
+
+    let entry = linesByImage.get(imageKey);
+    if (!entry) {
+      entry = { image, group: [] };
+      linesByImage.set(imageKey, entry);
+    }
+    entry.group.push({
+      image,
+      offset,
+      lineIndex,
+      parsedLine,
+    });
+  }
+
+  for (const { image, group } of linesByImage.values()) {
+    const { debugId, basename: moduleBasename, extname: moduleExtname } = image
+    const suffix = moduleExtname === '.pdb' ? '1' : '0';
+    const stream = await getSymbolFile(debugId.replace(/-/g, '') + suffix, moduleBasename)
+    if (!stream) {
+      continue;
+    }
+
+    const frames = group.map(({ offset }) => offset);
+    const symbolicated = await symbolicateFrames(stream, frames);
```

Comment:
In case you are wondering, this module streams once through the `.sym` file without building an index - you have to supply all the offsets that you want to identify right at the call time. Thus we have to group offsets by module first and then symbolicate them module-by-module.

On a good note, this is faster than building an AVL tree, and only an O(N) amount of memory where N is the number of frames.